controls/src/main/java/org/tweetwallfx/controls/stepengine/StepIterator.java

```diff
@@ -101,19 +104,48 @@ public Step next() {
 
     private static class Builder {
 
+        private static final Map<String, Step.Factory> FACTORIES;
+
+        static {
+            LOGGER.info("loading configurations data converters");
+            final Map<String, List<Step.Factory>> converters = StreamSupport
+                    .stream(ServiceLoader.load(Step.Factory.class).spliterator(), false)
+                    .peek(sf -> LOGGER.info("Registering Step.Factory '{}' which creates the Step '{}'", sf, sf.getStepClass().getCanonicalName()))
+                    .collect(Collectors.groupingBy(sf -> sf.getStepClass().getCanonicalName()));
+
+            // ensure there are no conflicting ConfigurationConverter registered for a specific key
+            converters.entrySet()
+                    .stream()
+                    .filter(e -> e.getValue().size() > 1)
+                    .findAny()
+                    .ifPresent(e -> {
+                        throw new IllegalArgumentException("At most one ConfigurationConverter may be registered to convert configuration data but the following ConfigurationConverters are registered: " + e.getValue());
+                    });
+
+            // only one element in the list. so use the one element in the list instead of the list
+            FACTORIES = converters.entrySet()
+                    .stream()
+                    .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().get(0)));
+        }
+
         private final List<Step> steps = new ArrayList<>();
 
-        public Builder addStep(final String classname) {
-            try {
-                Object newInstance = Thread.currentThread().getContextClassLoader().loadClass(classname).getDeclaredConstructor().newInstance();
-                if (newInstance instanceof Step) {
-                    steps.add((Step) newInstance);
+        public Builder addStep(final String stepClassName) {
+            final Step.Factory factory = FACTORIES.get(stepClassName);
+
+            if (null == factory) {
+                LOGGER.error("No Factory exist that can create Step '{}'", stepClassName);
+            } else {
+                final Step step = factory.create();
+
+                if (null == step) {

```

Comment:
Do we want to allow null as return of factory.create? Feels weird.

Reply:
I'm good with throwing an exception...